### PR TITLE
Fix #1159

### LINKED ISSUE #1159: ApplyLetRec arguments are not properly mapped in postconditions
The `inside` function is accepted by Stainless.

```scala
import stainless.lang._

object Test {

  def foo(): Unit = {
    var i = 0
    def isZero = i == 0

    def inside: Unit = {
      require(i <= 10)
      i += 1
    }.ensuring(_ =>
      isZero
    )

    inside
  }

}
```

The call `isZero` is transformed to `isZero(i)` where `i` refers to the initial value of `i` instead of the new value.

### PR DESCRIPTION
**Pull request by mario-bucev**
Following the suggestion https://github.com/epfl-lara/stainless/pull/1160#issuecomment-918938254

### CODE REVIEW

File: core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala [hunk 1]
```diff
@@ -242,46 +242,72 @@ class ImperativeCodeElimination(override val s: Trees)(override val t: s.type)
...
+                             (bodyWrapper: (Expr, Seq[Variable]) => Expr): (Expr, Expr => Expr, Seq[Variable]) = {
+                  assert(extraVarReplace.keySet.intersect(modifiedVars.toSet).isEmpty)
+                  val freshVarDecls: Seq[Variable] = modifiedVars.map(_.freshen)
+                  val rewritingMap: Map[Variable, Variable] = modifiedVars.zip(freshVarDecls).toMap
+
+                  val freshBody = postMap {
+                    case Assignment(v, e) => rewritingMap.get(v).map(nv => Assignment(nv, e))
+                    case v: Variable => rewritingMap.get(v).orElse(extraVarReplace.get(v))
+                    case Old(v: Variable) if freshVarDecls.contains(v) =>
+                      Some(freshVars(freshVarDecls.indexOf(v)))
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala:260]: Ideally, this should be more restrictive and only apply within the postcondition to allow the ImperativeCleanup phase to correctly catch all incorrect uses of `old`. Could you maybe make this part of the `extraVarReplace` (_e.g._ by making it into a `Map[Expr, Expr]`)?

**Review by samarion — APPROVED** [review]

File: core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala (hunk 1 above)
Comment (mario-bucev) [reply core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala:260]: Good point!

Comment (mario-bucev) [reply core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala:260]: (I'll push an updated version)

Comment (mario-bucev) [reply core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala]: Actually, the `Old(v)` expression will already have its variable rewritten to its freshened variant (contained in `freshVarDecls`). I pushed a version that is simply passed a flag to tell us whether we should rewrite `old` expressions (less elegant, that's for sure...).

**Comment by vkuncak** [comment]
@samarion again it's marked "approved". Did you mean to just comment without clicking "approve"?

**Pull request merged**